src/sinks/socket.rs

```diff
@@ -264,14 +265,22 @@ mod test {
                     future::poll_fn(move |cx| loop {
                         if let Some(fut) = close_rx.as_mut() {
                             if let Poll::Ready(()) = fut.poll_unpin(cx) {
-                                stream.get_ref().unwrap().shutdown(Shutdown::Write).unwrap();
+                                // TODO: Figure out a way not to block the thread here.
+                                // Spawning the future with `tokio::spawn` didn't work because the lifetime requirements
+                                // for the reference to the stream could not be met.
+                                tokio::runtime::Runtime::new()
+                                    .unwrap()
+                                    .block_on(stream.shutdown())
+                                    .unwrap();
```

Comment:
Need to fix that by introducing a state machine and polling the shutdown. Otherwise this will fail spectacularily if only one thread is available.

Reply:
Fixed in c74a65b.

Reply:
Fixed in 189caa5cad16dbc269efb4cab3f2b8872b8046f5, also see https://github.com/timberio/vector/pull/5872#discussion_r594130420.